"remove unnecessary condition" caused
```
sbt.ForkMain$ForkError: java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at scala.collection.immutable.ArraySeqTest.t10851(ArraySeqTest.scala:58)
```

--- a/src/library/scala/Array.scala
+++ b/src/library/scala/Array.scala
@@ -55,24 +55,9 @@ object Array {
    */
   def newBuilder[T](implicit t: ClassTag[T]): ArrayBuilder[T] = ArrayBuilder.make[T](t)
 
-  def from[A : ClassTag](it: IterableOnce[A]): Array[A] = {
-    val n = it.knownSize
-    if (n > -1) {
-      val elements = new Array[A](n)
-      val iterator = it.iterator
-      var i = 0
-      while (i < n) {
-        ScalaRunTime.array_update(elements, i, iterator.next())
-        i = i + 1
-      }
-      elements
-    } else {
-      val b = ArrayBuilder.make[A]
-      val iterator = it.iterator
-      while (iterator.hasNext)
-        b += iterator.next()
-      b.result()
-    }
+  def from[A : ClassTag](it: IterableOnce[A]): Array[A] = it match {
+    case it: Iterable[A] => it.toArray[A]
+    case _ => it.iterator.toArray[A]
   }
 
   private def slowcopy(src : AnyRef,

--- a/src/library/scala/collection/IterableOnce.scala
+++ b/src/library/scala/collection/IterableOnce.scala
@@ -162,7 +162,7 @@ final class IterableOnceExtensionMethods[A](private val it: IterableOnce[A]) ext
   @deprecated("Use .iterator.toArray", "2.13.0")
   def toArray[B >: A: ClassTag]: Array[B] = it match {
     case it: Iterable[B] => it.toArray[B]
-    case _ => mutable.ArrayBuffer.from(it).toArray
+    case _ => it.iterator.toArray[B]
   }
 
   @deprecated("Use .iterator.to(List) instead", "2.13.0")
@@ -1275,14 +1275,17 @@ trait IterableOnceOps[+A, +CC[_], +C] extends Any { this: IterableOnce[A] =>
 
   @`inline` final def toBuffer[B >: A]: mutable.Buffer[B] = mutable.Buffer.from(this)
 
-  /** Convert collection to array. */
+  /** Convert collection to array.
+    *
+    * Implementation note: DO NOT call [[Array.from]] from this method.
+    */
   def toArray[B >: A: ClassTag]: Array[B] =
     if (knownSize >= 0) {
       val destination = new Array[B](knownSize)
       copyToArray(destination, 0)
       destination
     }
-    else mutable.ArrayBuffer.from(this).toArray[B]
+    else mutable.ArrayBuilder.make[B].addAll(this).result()
 
   // For internal use
   protected def reversed: Iterable[A] = {

--- a/src/library/scala/collection/immutable/ArraySeq.scala
+++ b/src/library/scala/collection/immutable/ArraySeq.scala
@@ -208,18 +208,9 @@ object ArraySeq extends StrictOptimizedClassTagSeqFactory[ArraySeq] { self =>
 
   def empty[A : ClassTag]: ArraySeq[A] = emptyImpl
 
-  def from[A : ClassTag](it: scala.collection.IterableOnce[A]): ArraySeq[A] = unsafeWrapArray {
-    val n = it.knownSize
-    if (n > -1) {
-      val elements = Array.ofDim[A](n)
-      val iterator = it.iterator
-      var i = 0
-      while (i < n) {
-        ScalaRunTime.array_update(elements, i, iterator.next())
-        i = i + 1
-      }
-      elements
-    } else ArrayBuffer.from(it).toArray
+  def from[A](it: scala.collection.IterableOnce[A])(implicit tag: ClassTag[A]): ArraySeq[A] = it match {
+    case as: ArraySeq[A] => as
+    case _ => unsafeWrapArray(Array.from[A](it))
   }
 
   def newBuilder[A : ClassTag]: Builder[A, ArraySeq[A]] =

--- a/test/junit/scala/collection/immutable/ArraySeqTest.scala
+++ b/test/junit/scala/collection/immutable/ArraySeqTest.scala
@@ -53,12 +53,6 @@ class ArraySeqTest {
   }
 
   @Test
-  def t10851(): Unit = {
-    val s1 = ArraySeq.untagged(1,2,3)
-    assertTrue(s1.unsafeArray.getClass == classOf[Array[AnyRef]])
-  }
-
-  @Test
   def t11187(): Unit = {
     assertEquals(ArraySeq(1, 2), ArraySeq(2, 1).sorted)
     assertEquals(ArraySeq(1, 2), ArraySeq(2, 1).sortBy(identity))

--- a/test/junit/scala/collection/immutable/ArraySeqTest.scala
+++ b/test/junit/scala/collection/immutable/ArraySeqTest.scala
@@ -95,4 +95,11 @@ class ArraySeqTest {
     val y = Array[Byte](10).toSeq
     assertEquals(x.hashCode(), y.hashCode())
   }
+
+  @Test
+  def from(): Unit = {
+    val as = ArraySeq("foo", "bar", "baz")
+    assert(ArraySeq.from(as) eq as)
+    assert(ArraySeq(as: _*) eq as)
+  }
 }

